### PR TITLE
Do not warm up searcher in engine constructor

### PR DESCRIPTION
**Pull request by dnhatn**
This PR is another prerequisite for #47186.

With this change, we won't warm up searchers until we externally refresh an engine. We explicitly [refresh](https://github.com/elastic/elasticsearch/blob/aa0eb006d2e42f4eec6ddea599b0222cd45a43d4/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1312) before allowing reading from a shard (i.e., move to post_recovery state) and during [resetting](https://github.com/elastic/elasticsearch/compare/master...dnhatn:delay-warming?expand=1#diff-49e1a1b834b522f4ae6997c5defe9eb0R3378). These guarantees that we have warmed up the engine before exposing the external searcher.

Relates #47186

### CODE REVIEW

**Review by jpountz — COMMENTED** [review]
The change looks correct to me, but I'm not familiar enough with this part of the code to reason about potential implications, hopefully Henning or Yannick will be able to help here.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -368,7 +362,25 @@ protected int getRefCount(ElasticsearchDirectoryReader reference) {
...
+    }
+
+    @Override
+    final boolean assertSearcherIsWarmedUp(String source, SearcherScope scope) {
+        if (scope == SearcherScope.EXTERNAL) {
+            switch (source) {
+                // these sources can access the external searchers before the engine is warmed up
+                case "segments":
+                case "segments_stats":
+                case "refresh_needed":
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: why is this the case? This is only called by `scheduledRefresh`, which requires the shard to be started or post_recovery. We explicitly refresh however before moving to post_recovery state (see postRecovery method).

**Review by ywelsch — COMMENTED** [review]
Left one comment, o.w. looking good

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Good catch :). Fixed in https://github.com/elastic/elasticsearch/pull/48605/commits/204faaae77f565cc7d3521e4dfa61b60fdb4f75f

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -370,7 +364,24 @@ protected int getRefCount(ElasticsearchDirectoryReader reference) {
...
+    @Override
+    final boolean assertSearcherIsWarmedUp(String source, SearcherScope scope) {
+        if (scope == SearcherScope.EXTERNAL) {
+            switch (source) {
+                // we can access segment_stats while a shard is still in the recovering state.
+                case "segments":
+                case "segments_stats":
+                    break;
+                default:
+                    assert externalReaderManager.isWarmedUp : "searcher was not warmed up yet for source[" + source + "]";
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: This assert seems to access the boolean unsafely. Not guarded by refreshLock here AFAICS.

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java [hunk 3]
```diff
@@ -670,6 +670,7 @@ public Searcher acquireSearcher(String source, SearcherScope scope) throws Engin
         }
         Releasable releasable = store::decRef;
         try {
+            assert assertSearcherIsWarmedUp(source, scope);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Maybe also assert that searcher is warmed up when shard moves to `STARTED` state?

**Review by henningandersen — APPROVED** [review]
LGTM.

Left two minor comments to consider.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Yeah, the comment is outdated with the assertion. I think we have happens-before relationships in IndexShard (in postRecovery and resetEngineToGlobalCheckpoint). Are you okay to leave it as is?

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Did you mean to expose this assertion to IndexShard?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 2 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Looks good, thanks for clarifying.

File: server/src/main/java/org/elasticsearch/index/engine/Engine.java (hunk 3 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/engine/Engine.java]: Not necessary, let us skip this. It would not catch the reset engine case anyway.

**Pull request merged**